New exercise: enumerating trees.

fpottier opened the pull request:
This exercise uses the `Seq` API whose implementation was the topic of the exercises entitled "Symbolic Sequences as Objects" and "Symbolic Sequences as Data".

yurug commented on exercises/enumerating_trees/descr.md:
```diff
@@ -0,0 +1,364 @@
+
+Thus, an enumeration `e` of type `'a enum` represents a set of elements of
+type `'a`, grouped by size. The function call `e(s)` returns the subset of
+the elements of size `s`.
```
At some point, it could be a good idea to say that the sequence has no repetition.

fpottier replied:
Ok.

yurug commented on exercises/enumerating_trees/descr.md:
```diff
@@ -0,0 +1,364 @@
+**Question 7.** Define an enumeration `bit` of type `int enum`
+whose values, `0` and `1`, are both considered to have size 0.
+
+We assume that a memoizing fixed point combinator `fix` is given. Its type is
```
memoizing -> [memoizing](https://en.wikipedia.org/wiki/Memoization)

fpottier replied:
Ok.

yurug commented on exercises/enumerating_trees/descr.md:
```diff
@@ -0,0 +1,364 @@
+
+**Question 8.** Define a function `list` of type `'a enum -> 'a list enum` such
+that, if `e` is an enumeration of a certain set of elements, then `list e` is
+an enumeration of the lists of elements drawn from `e`. By convention, the
```
`list` must force the evaluation of `e 0; e 1; ...etc`, right? Can you motivate the need for this function? It seems inconsistent with the previous functions that are lazy.

fpottier replied:
I don't see what you mean. Calling `list e` does not do much computation; it only allocates a closure, say, `l`, which it returns. Then, when `l` is applied to some size `s`, then probably this will cause `e i` to be invoked for every `i` less than or equal to `s`. It is probably a good thing if the function `e` is memoized, as `e i` can be called more than once for a given `i`.

yurug replied:
Ignore this comment: I read `a enum list` instead of `a list enum`.

yurug commented on exercises/enumerating_trees/descr.md:
```diff
@@ -0,0 +1,364 @@
+`just x` represents a singleton set, whose element is `x`, and the size of
+this element is considered to be zero.
+
+**Question 3.** Define a function `pay` of type `'a enum -> 'a enum` such that
```
Can you give a bit of context to explain the naming choice for this function?

fpottier replied:
Ok. I have added the following:
*Note*. `pay` is so named because the elements of the enumeration `pay e` are
*more expensive*, so to speak, than the elements of the enumeration `e`.

yurug replied:
Did you push these changes?

fpottier replied:
Nope, as I said, I am having trouble because the repository upon which this PR was based has disappeared.

yurug commented:
Great exercise!

fpottier commented:
I have addressed your comments in my private repository, but I am having trouble updating this PR because its original repository has disappeared. If you could just accept this PR I could then commit my small changes afterwards.
